apps/ejabberd/src/ejabberd_auth_riak.erl

```diff
@@ -0,0 +1,269 @@
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+-module(ejabberd_auth_riak).
+
+-behaviour(ejabberd_gen_auth).
+
+-include("ejabberd.hrl").
+
+%% API
+-export([start/1,
+         stop/1,
+         store_type/1,
+         login/2,
+         set_password/3,
+         check_password/3,
+         check_password/5,
+         try_register/3,
+         dirty_get_registered_users/0,
+         get_vh_registered_users/1,
+         get_vh_registered_users/2,
+         get_vh_registered_users_number/1,
+         get_vh_registered_users_number/2,
+         get_password/2,
+         get_password_s/2,
+         get_password/3,
+         does_user_exist/2,
+         remove_user/2,
+         remove_user/3,
+         plain_password_required/0]).
+
+-export([bucket_type/1]).
+
+-spec start(ejabberd:lserver()) -> ok.
+start(_Host) ->
+    ok.
+
+-spec stop(ejabberd:lserver()) -> ok.
+stop(_Host) ->
+    ok.
+
+-spec store_type(ejabberd:lserver()) -> plain | scram.
+store_type(Host) ->
+    case scram:enabled(Host) of
+        false -> plain;
+        true -> scram
+    end.
+
+-spec set_password(ejabberd:luser(),ejabberd:lserver(), binary())
+        -> ok | {error, not_allowed | invalid_jid}.
+set_password(LUser, LServer, Password) ->
+    case prepare_password(LServer, Password) of
+        false ->
+            {error, invalid_password};
+        Password ->
+            User = mongoose_riak:fetch_type(bucket_type(LServer), LUser),
+            do_set_password(User, LUser, LServer, Password)
+    end.
+
+-spec check_password(ejabberd:luser(), ejabberd:lserver(), binary()) -> boolean().
+check_password(LUser, LServer, Password) ->
+    case do_get_password(LUser, LServer) of
+        false ->
+            false;
+        #scram{} = Scram ->
+            scram:check_password(Password, Scram);
+        Password when is_binary(Password) ->
+            Password /= <<"">>;
+        _ ->
+            false
+    end.
+
+-spec check_password(ejabberd:luser(),
+                     ejabberd:lserver(),
+                     binary(),
+                     binary(),
+                     fun()) -> boolean().
+check_password(LUser, LServer, Password, Digest, DigestGen) ->
+    case do_get_password(LUser, LServer) of
+        false ->
+            false;
+        #scram{} = Scram ->
+            scram:check_digest(Scram, Digest, DigestGen, Password);
+        PassRiak when is_binary(PassRiak) ->
+            ejabberd_auth:check_digest(Digest, DigestGen, Password, PassRiak)
```

Comment:
The check_password/3 function has  `_` clause.  Do we need the same clause here?